client/post-editor/editor-featured-image/dropzone.jsx

```diff
@@ -0,0 +1,114 @@
+/**
+ * External dependencies
+ */
+import React, { Component, PropTypes } from 'react';
+import { connect } from 'react-redux';
+import { head, uniqueId } from 'lodash';
+import { localize } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import DropZone from 'components/drop-zone';
+import MediaActions from 'lib/media/actions';
+import MediaStore from 'lib/media/store';
+import MediaUtils from 'lib/media/utils';
+import PostActions from 'lib/posts/actions';
+
+import { receiveMedia, deleteMedia } from 'state/media/actions';
+import { editPost } from 'state/posts/actions';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { getEditorPostId } from 'state/ui/editor/selectors';
+
+class FeaturedImageDropZone extends Component {
+	static propTypes = {
+		site: PropTypes.object.isRequired,
+		post: PropTypes.object.isRequired,
+	};
+
+	static contextTypes = {
+		store: PropTypes.object
+	};
+
+	onFilesDrop = ( files ) => {
+		/**
+		 * Filter files for `image` media prefix and return the first image.
+		 *
+		 * At the moment we ignore all the other images that were dragged onto the DropZone
+		 */
+		const droppedImage = head( MediaUtils.filterItemsByMimePrefix( files, 'image' ) );
+
+		if ( ! droppedImage ) {
+			return false;
+		}
+
+		const transientMediaId = uniqueId( 'featured-image' );
+		const siteID = this.props.siteId;
+
+		const handleFeaturedImageUpload = () => {
+			const media = MediaStore.get( siteID, transientMediaId );
+			const isUploadInProgress = media && MediaUtils.isItemBeingUploaded( media );
+			const isFailedUpload = ! media;
+
+			if ( isFailedUpload ) {
+				this.props.deleteMedia( siteID, transientMediaId );
+			} else {
+				this.props.receiveMedia( siteID, media );
+			}
+
+			/**
+			 * File uploaded successfully, there is no longer need to listen for the changes.
+			 */
+			if ( media && ! isUploadInProgress ) {
+				MediaStore.off( 'change', handleFeaturedImageUpload );
+			}
```

Comment:
Good catch. I updated the code to always unsubscribe when the media has been uploaded.